File: Meeting#2/java/src/WeightedQuickUnion.java
public class WeightedQuickUnion implements UnionFind {

    private final int[] id;
    private final int[] weight;
    private int count;

    public WeightedQuickUnion(final int size) {
        id = new int[size];
        for (int i = 0; i < size; i++) {
            id[i] = i;
        }
        weight = new int[size];
        for (int i = 0; i < size; i++) {
            weight[i] = 1;
        }
        count = size;
    }

    @Override
    public boolean connected(final int p, final int q) {
        return find(p) == find(q);
    }

    @Override
    public void union(final int p, final int q) {
        if (p == q) {
            return;
        }
        final int pRoot = find(p);
        final int qRoot = find(q);
        if (pRoot == qRoot) {
            return;
        }
        if (weight[pRoot] < weight[qRoot]) {
            id[pRoot] = qRoot;
            weight[qRoot] += weight[pRoot];
        } else {
            id[qRoot] = pRoot;
            weight[pRoot] += weight[qRoot];
        }
        count--;
    }

    @Override
    public int count() {
        return count;
    }

    @Override
    public int find(int p) {
        final int length = id.length;
        if (p < 0 || p >= length) {
            throw new IllegalArgumentException(String.format("P value %1$d isn't between 0 and %2$d", p, length));
        }
        return root(p);
    }

    private int root(int i) {
        while (i != id[i]) {
            i = id[i];
        }
        return i;
    }

}

File: Meeting#2/python/weighted_quick_union.py
class WeightedQuickUnion:

    def __init__(self, size):
        self._ids = list(range(0, size))
        self._weights = [1 for _ in range(size)]
        self._count = size
        pass

    def connected(self, p: int, q: int) -> bool:
        return self.find(p) == self.find(q)

    def union(self, p: int, q: int) -> None:
        if p == q:
            return
        p_root = self.find(p)
        q_root = self.find(q)
        if p_root == q_root:
            return
        if self._weights[p_root] < self._weights[q_root]:
            self._ids[p_root] = q_root
            self._weights[q_root] += self._weights[p_root]
        else:
            self._ids[q_root] = p_root
            self._weights[p_root] += self._weights[q_root]
        self._count -= 1

    def count(self) -> int:
        return self._count

    def find(self, p: int) -> int:
        if p < 0 or p >= len(self._ids):
            raise ValueError(f"P value {p} isn't between 0 and {len(self._ids)}")
        return self._root(p)

    def _root(self, i: int) -> int:
        while i != self._ids[i]:
            i = self._ids[i]
        return i
